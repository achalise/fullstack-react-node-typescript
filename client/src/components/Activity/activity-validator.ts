export const validate = (field: Field): ValidationResult => {
    let res = {status: true} as ValidationResult;
    switch (field.name) {
        case 'name':
            res = validateName(field);
            break;
        case 'type':
            res = validateType(field);    
            break;
        default:

    }
    return res;
}

const validateName = (field: Field) => {
    let res = {status: true} as ValidationResult;
    let value = field.value;
    if (!value || value.length < 3) {
        res.status = false;
        res.message = `Name must be at least 3 characters`;
    } 
    return res;
}

const validateType = (field: Field) => {
    let res = {status: true} as ValidationResult;
    let value = field.value;
    if (!value || value.length < 3) {
        res.status = false;
        res.message = `Password must be at least 5 characters`;
    } 
    return res;
}

export interface Field {
    name: string;
    value: string;
}

export interface ValidationResult {
    status: boolean;
    message?: string;
}

