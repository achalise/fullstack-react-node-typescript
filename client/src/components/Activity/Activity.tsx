import React from 'react';
import * as actions from '../../actions';
import { useFormValidator } from '../../form-utils/form-validator';
import { validate } from './activity-validator';
import { ErrorMessage } from '../ErrorMessage';
import { connect } from 'react-redux';
import { IActivity } from '../../domain/activity';

const InitialState = {
  name: '',
  type: '',
} as {
  name: string;
  type: string;
};

const Activity: React.FC<ActivityProps> = (props: ActivityProps) => {
  const [values, handleChange] = useFormValidator(InitialState, validate);
  const onClick = () => {
    console.log(props);
    console.log(`Submitting the form`, values);
    props.createActivity({type: values.type, name: values.name} as IActivity);
    props.history.push('/activities');
  };
  return (
    <div className="mt-5">
      <form>
        <div className="form-group row">
          <label htmlFor="name" className="col-sm-2 col-form-label">
            Name:
          </label>
          <div className="col-sm-10">
            <input
              id="name"
              type="text"
              name="name"
              className={'form-control'}
              onChange={handleChange}
              value={values.name}
            />
            {values.error && <ErrorMessage {...values.error['name']} />}
          </div>
        </div>
        <div className="form-group row">
          <label htmlFor="type" className="col-sm-2 col-form-label">
            Type:
          </label>
          <div className="col-sm-10">
            <input
              id="type"
              type="text"
              name="type"
              className="form-control"
              onChange={handleChange}
              value={values.type}
            />
            {values.error && <ErrorMessage {...values.error['type']} />}
          </div>
        </div>
        <div className="form-group row">
          <div className="col-sm-12 d-flex justify-content-end">
            <input type="button" className="btn-sm btn-primary mr-2" value="Cancel" />
            <input type="button" className="btn-sm btn-primary" onClick={onClick} value="Submit" />
          </div>
        </div>
      </form>
    </div>
  );
};

export default connect(
  null,
  actions,
)(Activity);

interface ActivityProps {
  history?: any;
  createActivity: (activity: IActivity) => {};
}
