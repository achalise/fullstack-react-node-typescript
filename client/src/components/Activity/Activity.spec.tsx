import React from 'react';
import {render, cleanup, waitForElement, wait, fireEvent, getByTestId, queryByAttribute} from '@testing-library/react'
import { Activity } from './Activity';

afterEach(cleanup);

describe(`Activity tests`, () => {
    it(`Should correctly load the activity entry form`, async (done) => {
        const {container, getByText} = render(<Activity/>);
        let element = await waitForElement(() => getByText('Name:'));
        expect(element).toMatchSnapshot();

        element = await waitForElement(() => getByText('Type:'));
        expect(element).toMatchSnapshot();
        done();
    })

    it(`Should correctly validate the text input field for name`, async(done) => {
        const byId = queryByAttribute.bind(null, 'id');
        const {container, getByText} = render(<Activity/>);
        let element = await waitForElement(() => byId(container, 'name'));
        expect(element).toMatchSnapshot();
        fireEvent.change(element, {
            target: {
                value: `as`
            }
        });
        let error = await waitForElement(() => getByText(`Name must be at least 3 characters`));
        expect(error).toMatchSnapshot();
        done();
    })
})
