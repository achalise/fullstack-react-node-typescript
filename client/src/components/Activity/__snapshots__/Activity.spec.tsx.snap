// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity tests Should correctly load the activity entry form 1`] = `
<label
  class="col-sm-2 col-form-label"
  for="name"
>
  Name:
</label>
`;

exports[`Activity tests Should correctly load the activity entry form 2`] = `
<label
  class="col-sm-2 col-form-label"
  for="type"
>
  Type:
</label>
`;

exports[`Activity tests Should correctly validate the text input field for name 1`] = `
<input
  class="form-control"
  id="name"
  name="name"
  type="text"
  value=""
/>
`;

exports[`Activity tests Should correctly validate the text input field for name 2`] = `
<div
  class="error-message"
>
  Name must be at least 3 characters
</div>
`;
